Introduce a migration doc for globus-sdk and globus-cli usage replacing this library (#201)
globus/globus-automate-client: docs/source/migration.rst

different parsers may treat identical documents differently.
``globus-cli`` and ``globus-sdk`` do not support YAML parsing, but it is possible
to convert YAML to JSON using a variety of tools.
This approach ensures that Globus provided software operates consistently, and
allows users to continue using YAML or to move off of it, as they prefer.

In this section, we will cover two popular tools for converting YAML to JSON,
yq (written in Go) and remarshal (written in Python). We will also cover
one python library, ``pyyaml``, which can be used to load YAML data and pass it
to the ``globus-sdk``.
Various other tools provide similar functionality in other languages, and there
are alternative parsers available in python.

remarshal
~~~~~~~~~

The `remarshal <https://github.com/remarshal-project/remarshal>`_ project
provides a wide range of commands for converting data between different
formats, including YAML and JSON.

These commands exist for the sole purpose of converting data between formats,
and are therefore a perfect fit for our use-case.

As ``remarshal`` is a python CLI, installation should be performed with
``pipx``, as with the ``globus-cli``.

> For full instructions, follow `remsarhsal's installation documentation

Reviewer: ```suggestion
For full instructions, follow `remarshal's installation documentation
```

Author: I can usually explain my topos as simple innversions of double characacters.
I dddon't have an essay explain nation four remsarhsal.

Migration to globus-sdk and globus-cli
======================================

> The Globus Automate Client has been deprecated.

Reviewer: ```suggestion
The Globus Automate Client is deprecated.
```

to convert YAML to JSON using a variety of tools.
This approach ensures that Globus provided software operates consistently, and
allows users to continue using YAML or to move off of it, as they prefer.

In this section, we will cover two popular tools for converting YAML to JSON,
yq (written in Go) and remarshal (written in Python). We will also cover
one python library, ``pyyaml``, which can be used to load YAML data and pass it
to the ``globus-sdk``.
Various other tools provide similar functionality in other languages, and there
are alternative parsers available in python.

yq
~~

The `yq <https://mikefarah.gitbook.io/yq/>`_ tool is a CLI utility similar to the
popular ``jq`` command.
It provides a wide variety of commands for manipulating and extracting data
from YAML documents.

Installation
++++++++++++

``yq`` is distributed for multiple platforms as a binary as well as Homebrew
and Snapcraft.
For these latter two, use:

.. code-block:: bash

    # macOS, brew
    brew install yq

    # ubuntu-based linux, snap
    snap install yq

To get a ``yq`` binary directly, or to use the docker based distribution,
follow

> `yq's installation instructions <https://github.com/mikefarah/yq/#install>`_.

Reviewer: I recommend that we avoid creating our own install instructions for `yq`. Instead, I recommend only linking to the project's own install instructions.

Author: TBH, `yq`'s documentation was what pushed me to write something here. I think those docs are quite unfriendly to novice users because the ordering is... strange.

Normally I would prefer not to reproduce someone else's docs, but in this case, I'm pretty concerned about throwing a less technical user straight at the `yq` repo.

Author: Rereading things, I'm feeling more moderate about this. I don't love that the `yq` docs put the easiest, most user-friendly options at the bottom, but it's not really _that_ bad. We'll see if anyone complains -- I suspect that nobody will.

Usage
+++++

In order to convert a flow from YAML to JSON using ``yq``, all that is needed
is a command which loads the YAML document and then outputs it as JSON.

> .. code-block:: bash
> 
>     yq -o=json foo.yaml > foo.json

Reviewer: `bash` works here, but it might be more appropriate to highlight this (and, I think, most of the examples here) as `console`.

https://pygments.org/docs/terminal-sessions/

```suggestion
.. code-block:: console

    $ yq -o=json foo.yaml > foo.json
```

Author: Good call! `console` is much better than `bash` for this.

remarshal
~~~~~~~~~

The `remarshal <https://github.com/remarshal-project/remarshal>`_ project
provides a wide range of commands for converting data between different
formats, including YAML and JSON.

These commands exist for the sole purpose of converting data between formats,
and are therefore a perfect fit for our use-case.

Installation
++++++++++++

As ``remarshal`` is a python CLI, installation should be performed with
``pipx``, as with the ``globus-cli``.

First `follow the pipx installation documentation
<https://pypa.github.io/pipx/installation/>`_ to ensure you have ``pipx``
installed.

Next, run the install command:

.. code-block:: bash

>     pipx install remarshal

Author: Remarshal's instructions are pretty good. Maybe I just put remarshal first and worry less about `yq`. I'm sure I included it here just for a parallel construction.

Usage
+++++

Of the many commands provided by ``remarshal``, the one we want is simply
``yaml2json``. After installing, all that is needed is to run:

.. code-block:: bash

    yaml2json foo.yaml foo.json

pyyaml
~~~~~~

> Unlike the previous two tools, ``pyyaml`` is a python library, not a CLI.

Reviewer: ```suggestion
Unlike the previous two tools, ``pyyaml`` is a Python library, not a CLI.
```

If you have a YAML flow definition and want to use it with the ``globus-sdk``,
you must parse it from YAML yourself and proivde it as a dictionary.

Installation
++++++++++++

``pyyaml`` can be installed with ``pip install pyyaml``.

Usage
+++++

``pyyaml`` provides the ``yaml`` package.
To parse a YAML file, ``foo.yaml``, into a python data structure, import it and

> use the ``load`` function::
> 
>     import yaml

Reviewer: This prevents syntax highlighting.

```suggestion
use the ``load`` function:

..  code-block:: python

    import yaml
```

Author: Ah, interesting. I thought that sphinx rendered unadorned codeblocks (`::`) as python blocks. Good to know so that I don't do that again.

Reviewer: Oh! It might. But I don't recommend relying on that in the general case, as this could be viewed here on GitHub, too. :wink:

    with open("foo.yaml") as fp:

>         data = yaml.load(fp)

Reviewer: ```suggestion
        data = yaml.safe_load(fp)
```

    # a check may be a wise precaution, as YAML documents can contain lists or
    # other non-dict data
    if not isinstance(data, dict):
        raise ValueError(
            f"YAML document 'foo.yaml' contained unexpected type {type(data)}"

>         )

Reviewer: I don't think that this belongs here.

Author: Fair enough; I'll remove it.

``globus-automate`` CLI commands and the new ``globus-cli`` commands.

This section provides more detailed guidance for converting commands between
the two, for commands and usages where the mapping is non-obvious.

Required Options vs Positional Arguments
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In general, the ``globus-cli`` uses positional arguments for all required
data, whereas the ``globus-automate`` CLI used required options in some cases.

The conversion is typically straightforward, requiring first that you read the
``globus-cli`` helptext and then order arguments appropriately if necessary.

For example, ``globus-automate flow deploy`` has been replaced with
``globus flows create``. Starting from an original command like so:

.. code-block:: bash

    $ globus-automate flow deploy --description bar --title foo --definition foo.json

The first step is to determine which CLI options are required and in what
order. Run ``globus flows create --help`` to see the help text:

.. code-block:: bash

    $ globus flows create --help
    Usage: globus flows create [OPTIONS] TITLE DEFINITION

      Create a new flow.

    # more text follows

>     ...

Author: I'll do a self-review pass after I apply suggestions for this.

With this information, we can see that ``TITLE`` is the first positional
argument and ``DEFINITION`` is the second. ``--description`` is still an
option.

The final command is therefore:

.. code-block:: bash

    globus flows create foo foo.json --description bar

Pagination Options
~~~~~~~~~~~~~~~~~~

A number of ``globus-automate`` commands provide options for paging through
data, typically ``--marker`` and ``--per-page``.
In ``globus-cli``, these options are replaced with a single option ``--limit``,
which controls the total number of results returned.

Under ``globus-automate``, users had precise control over pagination, while
under ``globus-cli`` all pagination is implicitly handled for the user.

> The two implementations trade off between simplicity for users vs fine-grained

Reviewer: ```suggestion
The two implementations trade off between simplicity for users vs. fine-grained
```

```suggestion
The two implementations trade off between simplicity for users versus fine-grained
```

loop.

For example, it can be scripted like so:

.. code-block:: bash

    #!/bin/bash

    RUN_ID="$1"
    echo "Poll until '$RUN_ID' terminates"

    NUM_TRIES=10
    until [ "$NUM_TRIES" -eq 0 ]; do
      status="$(globus flows run show "$RUN_ID" --jmespath "status" --format unix)"
      case "$status" in
        SUCCEEDED)
          echo "succeeded"
          exit 0
          ;;
        FAILED)
          echo "failed"
          exit 1
          ;;
        *)
          NUM_TRIES=$((NUM_TRIES - 1))
          sleep 30
          ;;
      esac
    done

    echo "Run '$RUN_ID' did not terminate after 10 tries"
    exit 3

``globus-automate flow run --watch``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This ``--watch`` flag is another instance of the same behavior described above.
Users needing to poll on run status can use ``globus flows run show`` as in the
preceding example.

.. _migrate_create_flows_client:

SDK Migration and ``create_flows_client``
-----------------------------------------

The ``create_flows_client`` helper has no singular replacement.

Instead, users should expect to write a small block of code to correctly

> authenticate pass the resulting authorizer to the matching client clas.  See

Reviewer: ```suggestion
authenticate and pass the resulting authorizer to the matching client class.  See
```

`the globus-sdk example usage
<https://globus-sdk-python.readthedocs.io/en/stable/examples/create_and_run_flow/>`_ for
an example of how to do this.

Why was this removed?
~~~~~~~~~~~~~~~~~~~~~

The ``create_flows_client`` helper attempts to consolidate functionality across
a disparate set of concerns.
However, implementers attempting to build applications on top of the Globus

> Flows API need finer grained control than could be provided through this

Reviewer: ```suggestion
Flows API need finer-grained control than could be provided through this
```
